mesh/mesh.cpp

```diff
@@ -1581,6 +1581,116 @@ void Mesh::SetAttributes()
    }
 }
 
+void Mesh::GetAttributeSetNames(std::set<std::string> &names) const
+{
+   names.clear();
+
+   for (auto const &attr_set : attr_sets)
+   {
+      names.insert(attr_set.first);
+   }
+}
+
+void Mesh::GetBdrAttributeSetNames(std::set<std::string> &names) const
+{
+   names.clear();
+
+   for (auto const &bdr_attr_set : bdr_attr_sets)
+   {
+      names.insert(bdr_attr_set.first);
+   }
+}
+
+
+void Mesh::SetAttributeSet(const std::string &set_name, const Array<int> &attr)
+{
+   attr_sets[set_name] = attr;
+}
+
+void Mesh::SetBdrAttributeSet(const std::string &set_name,
+                              const Array<int> &attr)
+{
+   bdr_attr_sets[set_name] = attr;
+}
+
+void Mesh::ClearAttributeSet(const std::string &set_name)
+{
+   attr_sets.erase(set_name);
+}
+
+void Mesh::ClearBdrAttributeSet(const std::string &set_name)
+{
+   bdr_attr_sets.erase(set_name);
+}
+
+void Mesh::AddToAttributeSet(const std::string &set_name, int attr)
+{
+   attr_sets[set_name].Append(attr);
+   attr_sets[set_name].Sort();
+   attr_sets[set_name].Unique();
+}
+
+void Mesh::AddToAttributeSet(const std::string &set_name,
+                             const Array<int> &attr)
+{
+   attr_sets[set_name].Append(attr);
+   attr_sets[set_name].Sort();
+   attr_sets[set_name].Unique();
+}
+
+void Mesh::AddToBdrAttributeSet(const std::string &set_name, int attr)
+{
+   bdr_attr_sets[set_name].Append(attr);
+   bdr_attr_sets[set_name].Sort();
+   bdr_attr_sets[set_name].Unique();
+}
+
+void Mesh::AddToBdrAttributeSet(const std::string &set_name,
+                                const Array<int> &attr)
+{
+   bdr_attr_sets[set_name].Append(attr);
+   bdr_attr_sets[set_name].Sort();
+   bdr_attr_sets[set_name].Unique();
+}
+
+void Mesh::RemoveFromAttributeSet(const std::string &set_name, int attr)
+{
```

Comment:
I wonder if this could be replaced with:
```
{
    Array<int> &attrs = attr_sets[set_name];

    attrs.DeleteFirst(attr);
}
```

1) This would require that 
```
void Mesh::SetAttributeSet(const std::string &set_name, const Array<int> &attr)
void Mesh::SetBdrAttributeSet(const std::string &set_name,  const Array<int> &attr)
```
do actually call `Sort()` and `Unique()`. 

2) Ideally `Array<T>` should have a member function named `DeleteAll(const T &el)` which does "the right thing" based on whether an internal flag indicates if it  the array is sorted and/or unique. This would remove later the duplication for `Mesh::RemoveFromBdrAttributeSet`.

Reply:
Good idea. This has been done. I also added a warning message if the named sets don't exist (in just these two "RemoveFrom*" methods).